src/ringbuffer.rs

```diff
@@ -130,37 +133,46 @@ where
 		// reaches `Index::max_value` because we want a ringbuffer.
 		let next_index = self.end.wrapping_add(1);
 		if next_index == self.start {
-			// overwrite the oldest item in the FIFO ringbuffer
+			// queue presents as empty but is not
+			// --> overwrite the oldest item in the FIFO ringbuffer
 			self.start = self.start.wrapping_add(1);
-			self.end = next_index;
-		} else {
-			self.end = next_index;
 		}
+		self.end = next_index;
 	}
 
 	/// Push an item onto the front of the queue.
 	///
 	/// Equivalent to `push` if the queue is empty.
 	fn push_front(&mut self, item: I) {
-		if self.start == self.end {
-			<Self as RingBufferTrait<I>>::push(self, item);
+		if self.is_empty() {
+			// queue is empty --> regular push
+			self.push(item);
 			return;
 		}
 		let index = self.start.wrapping_sub(1);
 		Self::Map::insert(index, item);
 		self.start = index;
+		if self.start == self.end {
+			// queue presents as empty but is not
+			// --> overwrite the most recent item in the queue
+			self.end = self.end.wrapping_sub(1);
+		}
 	}
 
 	/// Pop an item from the start of the queue.
 	fn pop(&mut self) -> Option<I> {
-		if self.start == self.end {
+		if self.is_empty() {
 			return None;
 		}
 		let item = Self::Map::take(self.start);
 		self.start = self.start.wrapping_add(1);
 
 		item.into()
 	}
+
+	fn is_empty(&self) -> bool {
+		self.start == self.end
```

Comment:
Above changes LGTM, but are you still working on this particular line? :p

Reply:
Well I changed the code in a way that this condition should hold true.

Reply:
'k after re-reading I got it. Now `self.start == self.end` should never happen unless the queue is **actually** empty :)